### ISSUE TITLE
Installation paths of libraries

### ISSUE THREAD

**Issue by sagitter** [Cantera/cantera#318]
Hi all,

cantera-2.2.1 installs `libcantera.so` and all Python files under `/usr/lib`. This is not correct if I use x86_64 systems.

`scons` configuration that I'm using:

```
INFO: SCons is using the following Python interpreter: /usr/bin/python
Configuration variables read from 'cantera.conf' and command line:
    CXX = '/usr/bin/g++'
    CC = '/usr/bin/gcc'
    prefix = '/usr'
    python_package = 'full'
    python_cmd = '/usr/bin/python2'
    python_array_home = '/usr/lib64/python2.7/site-packages/numpy'
    python3_package = 'y'
    python3_cmd = '/usr/bin/python3'
    python3_array_home = '/usr/lib64/python3.4/site-packages/numpy'
    matlab_toolbox = 'n'
    f90_interface = 'n'
    FORTRAN = '/usr/bin/gfortran'
    FORTRANFLAGS = '-O2 -g -pipe -Wall -Werror=format-security ,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic'
    doxygen_docs = True
    sphinx_cmd = '/usr/bin/sphinx-build'
    use_sundials = 'y'
    sundials_include = '/usr/include/sundials'
    sundials_libdir = '/usr/lib64'
    blas_lapack_libs = 'lapack,blas,tatlas,satlas'
    blas_lapack_dir = '/usr/lib64'
    env_vars = 'all'
    cxx_flags = '-O2 -g -pipe -Wall -Werror=format-security ,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic'
    cc_flags = '-O2 -g -pipe -Wall -Werror=format-security ,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic'
    optimize = False
    optimize_flags = ''
    no_optimize_flags = ''
    debug = False
    debug_flags = ' -g -O0'
    warning_flags = ''
    extra_inc_dirs = '/usr/include/atlas-x86_64-base -I/usr/include/atlas'
    extra_lib_dirs = '/usr/lib64/atlas'
    build_thread_safe = True
    boost_inc_dir = '/usr/include'
    boost_lib_dir = '/usr/lib64'
    boost_windows_libs = ''
    F77 = '/usr/bin/gfortran'
    F77FLAGS = '-O2 -g -pipe -Wall -Werror=format-security ,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic'
    VERBOSE = True
```

**Comment by bryanwweber** [comment]
What distro are you using? My understanding is that certain distributions use `/usr/lib` or `/usr/local/lib` even for 64-bit libraries (e.g., Ubuntu does this IIRC; see: http://unix.stackexchange.com/q/43190/40481). Presumably, you expect it to be installed in `/usr/lib64` given the prefix you have?

**Comment by sagitter** [comment]
I'm compiling `cantera` on Fedora (I'm going to build it for official repositories)

> Presumably, you expect it to be installed in /usr/lib64 given the prefix you have?

Precisely.

**Comment by sagitter** [comment]
Can you change installation path?
Otherwise, please, let me know how to fix this behavior.

**Comment by bryanwweber** [comment]
Edited to add: I just noticed that you said you're going to be building this for the official repositories, and I want to note that my advice below should only be used if you completely control the machine you'll be installing onto. I don't think it will generalize, and I don't want to be distributing bad advice :smiley: If you're going to be building for repositories, its probably better to add a new option to the `layout` option, which can currently be `debian`, `compact`, or `standard` (the default). Maybe adding an option for `fedora`, presuming that most Fedora machines are similar, would be better.

---

One way to do this without editing any Cantera files may be to symlink the contents of `/usr/lib` into `/usr/lib64`. I'm not sure if that would introduce other problems though. Probably the easiest way to do this in a one-off fashion by editing the Cantera code is to edit the SConstruct file, at line 1113: https://github.com/Cantera/cantera/blob/2.2/SConstruct#L1113

I think changing the `lib` there to `lib64` will do what you want. Please note that I haven't tried this, I'm only making a guess. And also, this is a very hacky, temporary solution while we try to think of something more permanent.

@arghdos I think server 2 runs RHEL right? What have you done for that server?

**Comment by sagitter** [comment]
> I think changing the lib there to lib64 will do what you want. Please note that I haven't tried this, I'm only making a guess. And also, this is a very hacky, temporary solution while we try to think of something more permanent.

It's working.

> If you're going to be building for repositories, its probably better to add a new option to the layout option, which can currently be debian, compact, or standard (the default). Maybe adding an option for fedora, presuming that most Fedora machines are similar, would be better.

`rhel` would be perfect. Even if it's nearly like `debian`, library paths apart.

Where installation path of python files is set?
Currently, `cantera` script installs in `/usr/lib`:

```
  Python 2 package (cantera)  /home/sagitter/rpmbuild/BUILDROOT/cantera-2.2.1-1.fc23.x86_64/usr/lib/python2.7/site-packages
  Python 2 samples            /home/sagitter/rpmbuild/BUILDROOT/cantera-2.2.1-1.fc23.x86_64/usr/lib/python2.7/site-packages/cantera/examples 
  Python 3 package (cantera)  /home/sagitter/rpmbuild/BUILDROOT/cantera-2.2.1-1.fc23.x86_64/usr/lib/python3.4/site-packages
  Python 3 samples            /home/sagitter/rpmbuild/BUILDROOT/cantera-2.2.1-1.fc23.x86_64/usr/lib/python3.4/site-packages/cantera/examples 


Regards.
```

**Comment by speth** [comment]
Yes, that's the line of `SConstruct` that needs to be modified to get the library files in the right location.

The most general solution I guess would be to add a new configuration variable name such as `libdir_name` which can be `lib` by default or set to `lib64` by the user if they're compiling on Fedora or RHEL. Ideally, there would be some way to autodetect the correct default, but I'm not sure what that would be.

I'm not sure whether or not it makes sense to add a new `layout` option for building RPMs. The reason the `debian` layout exists is to divide things into the distinct sub-packages, i.e. cantera-common, cantera-dev, and cantera-python. If the same is useful for building RPMs, then adding a new layout makes sense, but otherwise I think just changing `ct_libdir` makes more sense.

Where are Python packages supposed to end up on RHEL/Fedora? Does it differ between packages that are managed by RPM versus packages that are installed manually. For Debian, this is handled by To the command line argument to `python setup.py install` of `--install-layout=deb`, which is controlled in the Cantera SCons scripts by the `PYTHON_INSTALLER` variable. I'm not sure what the expected arguments to `setup.py` are for RPM builds, though to get it to do the right thing.

**Comment by sagitter** [comment]
Hi speth,

> The reason the debian layout exists is to divide things into the distinct sub-packages, i.e. cantera-common, cantera-dev, and cantera-python. If the same is useful for building RPMs, then adding a new layout makes sense, but otherwise I think just changing ct_libdir makes more sense.

Indeed, that's what I will do. Package splitting (cantera, python2-cantera, python3-cantera) is expected in Fedora as well.

> Where are Python packages supposed to end up on RHEL/Fedora?

Where pure python2 modules are installed
`/usr/lib/python2.X/site-packages`

Where python2 extension modules that are compiled C are installed
`/usr/lib64/python2.X/site-packages` on x86_64
`/usr/lib/python2.X/site-packages` on x86

Where pure python3 modules are installed
`/usr/lib/python3.X/site-packages`

Where python3 extension modules that are compiled C are installed
`/usr/lib64/python3.X/site-packages` on x86_64
`/usr/lib/python3.X/site-packages` on x86

**Comment by speth** [comment]
If the goal is to create split packages, then definitely a new `layout=rpm` option should be introduced.

So currently the Cantera Python is being installed to `/usr/lib` even though it has compiled extension, right? We've had issues with this in the past where setup.py does not identify that there is a compiled extension (because it is compiled by SCons, not as part of setup.py). This is the reason for the patch to `bdist_msi.run` in setup.py.in. I don't know if something similar will be required here.

Before going any further on this particular front, does it make sense to use `bdist_rpm` to create the RPMs for the Python packages?

**Comment by sagitter** [comment]
> So currently the Cantera Python is being installed to /usr/lib even though it has compiled extension, right?

Yes.

> Before going any further on this particular front, does it make sense to use bdist_rpm to create the RPMs for the Python packages?

Personally, RPMs creation is managed by specific tools, therefore I won't use Python script as you propose. It could be an additional option.

**Comment by speth** [comment]
You're going to be using `setuptools` to put the Python package's files where they belong one way or another. `bdist_rpm` is just one of the options that can be specified. If RPM creation for Python packages is typically done with `bdist_rpm`, it makes sense to use it since it should know about the Fedora filesystem, but if not then perhaps whatever process is used for non-Python packages should be used instead. 

In either case, the Cantera build scripts aren't what is determining the problematic part of the path that the Python package is being installed to -- all the SCons scripts specify is that `prefix` is `/usr`. The rest is up to `setuptools`.

**Comment by speth** [comment]
I think the issues here should be resolved by #346. Running `scons install` with the `layout=rpm` will create a directory `rpm` with subdirectories for the files that should be in each of the sub-packages, e.g. `cantera-common`, `cantera-python`, `cantera-python3`, and `cantera-dev`.

@sagitter Can you confirm that these patches satisfy the needs to be able to generate working RPM packages?

**Comment by sagitter** [comment]
Is #346 a patch for [Cantera 2.2.1](https://github.com/Cantera/cantera/archive/v2.2.1.tar.gz) ?

**Comment by speth** [comment]
No, this is a patch for the development version (ultimately, 2.3.0)

**Comment by sagitter** [comment]
> I think the issues here should be resolved by #346. Running scons install with the layout=rpm will create a directory rpm with subdirectories for the files that should be in each of the sub-packages, e.g. cantera-common, cantera-python, cantera-python3, and cantera-dev.
> @sagitter Can you confirm that these patches satisfy the needs to be able to generate working RPM packages?
- It would be better if user can chose where install all files: a common destination directory instead of rpm/cantera-common, rpm/cantera-python, rpm/cantera-python3, rpm/cantera-dev like happens with Make for example.
- Python3 files are still installed in `/usr/lib` on 64bit system.
- The shared library `libcantera.so` is not linked against `sundials` libraries. Please, set a `soname` string if you can (`libcantera.so.2.3.0`).
- Cantera's Python2/3 libraries (`_cantera.so`) are not linked against `sundials` libraries and related Python library (`libpython2.X.so` and `libpython3.X.so`)
- None `Cantera` library is linked against `Boost`.

That's what I noted by compiling `Cantera-2.3.0a2` on Fedora 24 64bit.

Thanks.

**Comment by speth** [comment]
Latest updates to #346 (49d04a1) should fix the following:
- Python3 files are now installed to `/usr/lib64'
- `libcantera_shared.so` is now generated with sonames
- `libcantera_shared.so` now links to Sundials when needed

Regarding the other issues:
- There are lots of options for installation, but I don't know what you want, so you'll need to elaborate. Having the files that should be in each sub-package installed to a separate directory was useful when I was building Debian packages, and I presumed that this would be useful for RPM builds as well. If you want to install everything together, you can just use the `stage_dir` option to SCons instead of the `layout=rpm` option (and I can remove that option if it is not useful).
- If the Python shared libraries aren't linked to Sundials, then Cantera isn't using the system Sundials. Are you compiling with the option `system_sundials=y`?
- I don't know what "None Cantera library is linked against Boost." means. Cantera does not currently use any of the compiled Boost libraries (although older versions did).
- Linking to `libpython` is not necessary, since these libraries will only ever be loaded as Python extension modules.

**Comment by sagitter** [comment]
> Regarding the other issues:
> 
>   There are lots of options for installation, but I don't know what you want, so you'll need to elaborate. Having the files that should be in each sub-package installed to a separate directory was useful when I was building Debian packages, and I presumed that this would be useful for RPM builds as well. If you want to install everything together, you can just use the stage_dir option to SCons instead of the layout=rpm option (and I can remove that option if it is not useful).

`stage_dir` option does exactly what I meant.

>  If the Python shared libraries aren't linked to Sundials, then Cantera isn't using the system Sundials. Are you compiling with the option system_sundials=y?

`system_sundials=y` is set, whence I expect linking to Sundials.

>  I don't know what "None Cantera library is linked against Boost." means. Cantera does not currently use any of the compiled Boost libraries (although older versions did).

`scons` provides options to use `Boost` libraries ( `boost_inc_dir`); I didn't know this function was obsoleted.

**Comment by sagitter** [comment]
> Latest updates to #346 (49d04a1) should fix the following:

```
Python3 files are now installed to `/usr/lib64'
libcantera_shared.so is now generated with sonames
libcantera_shared.so now links to Sundials when needed
```

There is a little bug of `cantera.pc`: `libdir` is always `${prefix}/lib`

**Comment by speth** [comment]
changes in 6a33fd7:
- I have removed the `layout=rpm` option since that appears does not appear to be useful
- `cantera.pc` should now reference the correct directory

As far as the Sundials issue goes, can you provide the full output of a clean `scons build` (i.e. after having run `scons clean`)?

**Comment by sagitter** [comment]
> As far as the Sundials issue goes, can you provide the full output of a clean scons build (i.e. after having run scons clean)?

Linking is now correct:

```
$ ldd /usr/lib64/python2.7/site-packages/cantera/_cantera.so
    linux-vdso.so.1 (0x00007fff71fad000)
    libcantera.so.2 => /lib64/libcantera.so.2 (0x00007f9991866000)
    libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f99914df000)
    libm.so.6 => /lib64/libm.so.6 (0x00007f99911d4000)
    libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f9990fbd000)
    libc.so.6 => /lib64/libc.so.6 (0x00007f9990bfb000)
    libsundials_cvodes.so.2 => /lib64/libsundials_cvodes.so.2 (0x00007f99909c2000)
    libsundials_ida.so.2 => /lib64/libsundials_ida.so.2 (0x00007f99907a4000)
    libsundials_nvecserial.so.0 => /lib64/libsundials_nvecserial.so.0 (0x00007f99905a0000)
    libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f9990383000)
    /lib64/ld-linux-x86-64.so.2 (0x000056503a376000)
    liblapack.so.3 => /lib64/liblapack.so.3 (0x00007f998fb78000)
    libblas.so.3 => /lib64/libblas.so.3 (0x00007f998f921000)
    libgomp.so.1 => /lib64/libgomp.so.1 (0x00007f998f6f2000)
    libgfortran.so.3 => /lib64/libgfortran.so.3 (0x00007f998f3c1000)
    libdl.so.2 => /lib64/libdl.so.2 (0x00007f998f1bc000)
    libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00007f998ef7c000)

$ ldd /usr/lib64/python3.5/site-packages/cantera/_cantera.cpython-35m-x86_64-linux-gnu.so
    linux-vdso.so.1 (0x00007ffefd2ff000)
    libcantera.so.2 => /lib64/libcantera.so.2 (0x00007f8a17e30000)
    libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f8a17aa9000)
    libm.so.6 => /lib64/libm.so.6 (0x00007f8a1779e000)
    libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f8a17587000)
    libc.so.6 => /lib64/libc.so.6 (0x00007f8a171c5000)
    libsundials_cvodes.so.2 => /lib64/libsundials_cvodes.so.2 (0x00007f8a16f8c000)
    libsundials_ida.so.2 => /lib64/libsundials_ida.so.2 (0x00007f8a16d6e000)
    libsundials_nvecserial.so.0 => /lib64/libsundials_nvecserial.so.0 (0x00007f8a16b6a000)
    libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f8a1694d000)
    /lib64/ld-linux-x86-64.so.2 (0x000055f88d804000)
    liblapack.so.3 => /lib64/liblapack.so.3 (0x00007f8a16142000)
    libblas.so.3 => /lib64/libblas.so.3 (0x00007f8a15eeb000)
    libgomp.so.1 => /lib64/libgomp.so.1 (0x00007f8a15cbc000)
    libgfortran.so.3 => /lib64/libgfortran.so.3 (0x00007f8a1598b000)
    libdl.so.2 => /lib64/libdl.so.2 (0x00007f8a15786000)
    libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00007f8a15546000)
```

This is the build log from Fedora 24 64bit:
https://kojipkgs.fedoraproject.org//work/tasks/9068/14779068/build.log

Where these `undefined symbols`come from ?

```
cantera.x86_64: W: undefined-non-weak-symbol /usr/lib64/libcantera.so.2.3.0 fmt::internal::PrintfFormatter<char>::format(fmt::BasicWriter<char>&, fmt::BasicCStringRef<char>)
cantera.x86_64: W: undefined-non-weak-symbol /usr/lib64/libcantera.so.2.3.0 fmt::BasicFormatter<char>::format(fmt::BasicCStringRef<char>)
```

**Comment by speth** [comment]
The undefined symbols are from `fmt`. I'm expecting those to go away when using the `system_fmt=y` build option. Does that take care of everything?

**Comment by speth** [comment]
@sagitter Can you confirm that #346 resolves all of the issues that you have identified here? If there are any new problems, we can address those separately.

**Comment by sagitter** [comment]
> Can you confirm that #346 resolves all of the issues that you have identified here? If there are any new problems, we can address those separately.

I confirm.

**Comment by band-a-prend** [comment]
ping @speth 

Hello.

I don't know is it worth to create new issue or just ask here.

I try to write ebuild (build and installation script) for Gentoo Linux. The compilation and installation by package manager are made in sandbox, so if I want to install Cantera (or any package) to system "/usr" subfolders then I should  set 'prefix' to "${D%/}/usr" where  first part of the path "${D%/}" is the path within sandbox, i.e. "/var/tmp/portage/sci-libs/cantera-2.3.0/image" in my case. After that gentoo package manager moves this installed files to real system to "/usr".

I can initialy to set "prefix=/usr" to build with with option but when "scons install prefix="${D%/}/usr" is called to initially install cantera to appropriate sandbox directory then "install" script begin to build byte-code (.pyc-files) with substituting old prefix-path with new prefix-path ("/var/tmp/portage/sci-libs/cantera-2.3.0/image/usr"). Also some of 'plain-text' files are modified in similar way but it's not problem to replace appropriate strings with 'sed' after that before actual installation to system. But I don't know is it safety to replace strings within byte-code.

Is it possible to build byte-code .pyc files within "scons build" with other prefix ("/usr") and then call "scons install" with new "prefix="${D%/}/usr" without rebuilding of byte-code?

**Comment by bryanwweber** [comment]
Hi @band-a-prend This would probably be better as a new issue, and if you could include more detail of the procedure to build a Gentoo package, that would be very helpful (build scripts, links to documentation, etc.)

**Comment by band-a-prend** [comment]
Thanks a lot for reply. I re-read cantera configuration options and it seems I fixed Gentoo build script by adding `stage_dir="${D%/}"` configuration option that was added for Debian package. Where `${D%/}` - a variable that point to sandbox installation directory for Gentoo. And now default CANTERA_DATA and cantera libraries search into correct installation system paths.

I need some additional testing and if something go wrong then I'll create new issue.
